Add multicast push and improve TDBU control for Motion Blinds

starkillerOG opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
1) Add multicast push callbacks to instantly receive updates from the motion gateway and its connected blinds.

2) Add discovery to the config flow so users don't need to know the IP of the Motion Gateway (custom discovery protecol from upstream library, ssdp or mdns/zeroconf discovery is not yet supported by the firmware of the gateway, I am discussing with the manufacturers if ssdp or mdns/zeroconf can be added to the firmware, they are consulting their software developers).

3) Rework the Update function upstream to actually fetch the status data from the blind itself instead of only the cache of the gateway. Now communication goes like this: HomeAssistant -(wifi)-> motion gateway -(433MHz)-> blind -(433MHz)-> gateway -(wifi)-> HomeAssistant.
Before the communication was:  HomeAssistant -(wifi)-> motion gateway -(wifi)-> HomeAssistant
This ensures the positions cann't get out of sync and the battery status is actually fetched (was not the case before, it only refresehd when the app was opened which caused communication with the blind so the gateway cache would update).

4) Add improved control for Top Down Bottom Up (TDBU) motion blinds.
The sliders for postion are now scaled such that 0 and 100 cover exactly the allowed space in which the blind can move at the moment. (Bottom is able to move from 0 to the absolute position of the Top, and Top can move from the absolute postion of the Bottom to 100)
Also adds a "combined" entity for a TDBU blind, with that entity you can move Top and Bottom at the same time and shift the whole thing up or down.

5) Add motion_blinds.set_absolute_position service. For most blinds this does the same as cover.set_cover_position service, but for TDBU blinds it will set the absolute position relative to the window itself instead of the scaled position relative to the space in which the blind is allowed to move which cover.set_cover_position does for TDBU blinds.

Also include extra documentation for the TDBU blind.

Note that I wrote the motionblinds integration which was accepted only a few days ago.

Github comparison for upstream library changes for discovery: https://github.com/starkillerOG/motion-blinds/compare/0.3.0...0.4.7
Github comparison for upstream library changes for push callbacks: https://github.com/starkillerOG/motion-blinds/compare/0.2.1...0.3.0
Github comparison for upstream library changes for TDBU control: https://github.com/starkillerOG/motion-blinds/compare/d07477f8f0c0a73c474dfe5b385ad569af41c6b8..a5758229043642c34958ce8d6a9d4942cd47ad39

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

Config flow

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15756

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
Improve TDBU control and add combined control

starkillerOG committed:
Improve TDBU control and add combined control

frenck commented:
Thanks for bumping the dependency! However, could you please update the PR description to contain at least one (or multiple) of the following:

- A link to the release notes of this package version, and all versions in between.
- A link to the changelog of this package.
- A link to a Git(Hub) diff/compare view from the current version to the bumped version.

This allows us to review upstream changes, which is needed to decide this change is working as intended and/or if we can include it in, for example, a patch release of Home Assistant.

Thanks already! 👍

starkillerOG commented:
@frenck thanks, I just included a link.
Took me a while to figure out how to do this, from now on I will also release versions on Github instead of only on Pypi, that schould making these comparisions easier in the future.

starkillerOG committed:
fix black

starkillerOG commented:
Alright, I thinks this is ready for revieuw/merging.

starkillerOG committed:
Add multicast push

Reduce polling interval from 10 seconds to 15 minutes and add multicast push to instantly receive updates.

starkillerOG committed:
fix black and flake

starkillerOG committed:
fix config flow

starkillerOG committed:
update manufacturer name

as requested by the company itself.

starkillerOG committed:
fix import

starkillerOG committed:
Add set_absolute_position service for TDBU covers

This allows TDBU covers to set the Top, Bottom or Combined entity to an absolute position with respect to the window instead of the scaled reletive postion with respect to the available space in which the entity is allowed to move.

starkillerOG committed:
fix black

starkillerOG committed:
add option to set width

starkillerOG committed:
add service discription

starkillerOG committed:
fix styling

starkillerOG committed:
Add discovery to config flow

starkillerOG commented:
@bdraco would you be willing to revieuw this PR, since you revieuwed the initial PR for the motion_blinds component a week ago?

starkillerOG committed:
bump motionblinds to v0.4.2

Improve status update communication

starkillerOG commented:
@JPHutchins and @allenporter would you like testing out this code as well, since you revieuwed the intial Motion_blinds PR and I think you even have motion blinds?

starkillerOG committed:
Make sure there is intial data

starkillerOG committed:
prevent 1 blind from blocking update others

starkillerOG committed:
improve unavailble handling

starkillerOG committed:
fix double callback id

starkillerOG committed:
fix update TDBU

starkillerOG committed:
move available logic upstream

starkillerOG committed:
fix tests

starkillerOG committed:
fix tests

allenporter commented:
I actually can't test out multicast easily since my home assistant is on the public side of my wifi router nat and I set up a port forward to the gateway up to be able to access it.

starkillerOG commented:
@allenporter alright i see, would it be possible to also setup a port forward for multicast adress '238.0.0.18' port 32101?
I am actually not sure if something like that works for multicast...

allenporter commented:
I don't think my wifi router would support that.

For what it's worth, I just now went and tested motion blinds on 1.0.0b1 and it works great.  I used the IP of the gateway in setup, even through my hacky wifi router port forward thing.   I was able to see the cover positions, and issue calls to open, close, and set a specific position using the developer tools.

Maybe what I could do is just download motion blinds python library onto a laptop and try the discovery out there.  I made a little debug tool to try it out a few weeks ago to test out the port forward:
https://github.com/starkillerOG/motion-blinds/compare/main...allenporter:main which I can contribute back if you have any interest.

starkillerOG committed:
tweak timeouts

starkillerOG commented:
@allenporter I would be intrested in the debug tool, probably will need some small tweaks and cleanup, but I am willing to merge if you make a PR.

It might be a good idea to test out the newest version of motionblinds on your network, this newest version relays havily on the multicast, not only for instant pushes but together with the employees of Coullise B.V. I figured out that the current update process is only retrieving status from the cache of the gateway and not from the blinds directly.
I fixed that, but now also the update function uses multicast to receive the new data directly from the blind.

Therefore it might not work on your network if the multicast traffic gets blocked.
Would like to know if this is the case and see how we can resolve that.

starkillerOG commented:
@allenporter you can try this PR by using motion_blinds as a custom_component from here: https://github.com/starkillerOG/HA-motion-blinds
I keep that repositry updated (for now) so the employees at Coullise B.V. can test out the code on a whole range of products including some prototypes that are not yet on the market.

allenporter commented:
I tried to set this up using the ip address of the port forwarding w/ the gateway (given my odd network setup, couldn't test out multicast) and ran into this:

2020-12-08 06:34:00 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.
2020-12-08 06:34:25 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.
2020-12-08 06:34:50 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.

<img width="449" alt="Motion Blinds error" src="https://user-images.githubusercontent.com/6026418/101448958-c8566d80-38dc-11eb-91be-2b79f3b72bc3.png">

I haven't been able to try the real multicast solution yet.

allenporter commented on homeassistant/components/motion_blinds/cover.py:
```diff
@@ -146,6 +169,37 @@ def is_closed(self):
+    async def async_added_to_hass(self):
+        """Subscribe to multicast pushes and register signal handler."""
+        self._blind.Register_callback(self.unique_id, self.push_callback)
+        async_dispatcher_connect(self.hass, DOMAIN, self.signal_handler)
```
I think a best practice is to unregister this when the entity is removed
```suggestion
        self.async_on_remove(async_dispatcher_connect(self.hass, DOMAIN, self.signal_handler))
```

starkillerOG replied:
agreed

allenporter commented on homeassistant/components/motion_blinds/cover.py:
```diff
@@ -146,6 +169,37 @@ def is_closed(self):
         return self._blind.position == 100
 
+    @callback
+    def push_callback(self):
```
I believe this is only used by this class, so consider prefixing with "_"

starkillerOG replied:
good sugestion, done

allenporter commented on homeassistant/components/motion_blinds/__init__.py:
```diff
@@ -1,25 +1,71 @@
     """Set up the Motion Blinds component."""
+
+    def service_handler(service):
+        method = SERVICE_TO_METHOD.get(service.service)
```
service.service and method["method"] appear to be equivalent here.  Given that there is only one service at the moment, I might suggest unrolling this into a single hass.services.register call with a direct callback that implements handle_set_absolute_position then add the generality when adding a second service.

starkillerOG replied:
I would rather keep it generic like this, this is also how it is done in other integration such as the denonavr (that also only has 1 service at the moment). In this way it will be easy to add aditional services later on.

allenporter reviewed (commented):
One suggestion I have might be to break up the 5 different changes in the PR into separate smaller PRs that can be reviewed a bit faster, and independently.

starkillerOG commented:
> 
> 
> I tried to set this up using the ip address of the port forwarding w/ the gateway (given my odd network setup, couldn't test out multicast) and ran into this:
> 
> 2020-12-08 06:34:00 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.
> 2020-12-08 06:34:25 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.
> 2020-12-08 06:34:50 ERROR (SyncWorker_2) [motionblinds.motion_blinds] Timeout of 5.0 sec occurred on 5 attempts while waiting on multicast push from update request, communication between gateway and blind might be bad.
> 
> <img alt="Motion Blinds error" width="449" src="https://user-images.githubusercontent.com/6026418/101448958-c8566d80-38dc-11eb-91be-2b79f3b72bc3.png">
> 
> I have 3 blinds, so it seems like it was able to talk to the gateway to get the list of devices, but not able to query them directly.
> 
> When I tried out 1.0.0b1 i didn't have any problems using the gateway / cache for what its worth and was able to see current values reflected OK, though I did not test it out exhaustively.
> 
> I haven't been able to try the real multicast solution yet.
> 
> Thanks for your work on this integration.

This is the expected behavier, withouth the multicast it will generate these errors.
Let's see if you can get the port forwarding working for the multicast.

If that turns out to not be possible, I can see if I can make a setting that will silence these warnings. (that will mean you lose the available/unavailable tracking of the blinds and the true status updates will lack behind with 10 minutes, or at least will be assumed that the multicast has been processed by the gateway and the cache updated, basically you will lack some features, but I can make the basics work like in 1.0.0b1)

starkillerOG committed:
unregister when entity removed

Co-authored-by: Allen Porter <allen.porter@gmail.com>

starkillerOG committed:
protect _push_callback function

starkillerOG committed:
Merge branch 'patch-40' of https://github.com/starkillerOG/home-assistant into patch-40

starkillerOG committed:
fix black

starkillerOG commented:
> One suggestion I have might be to break up the 5 different changes in the PR into separate smaller PRs that can be reviewed a bit faster, and independently.

I had thought of that, but the diffrent parts relay on diffrent versions of the upstream library and have verious other small bug fixes in them, so I think it will get a bit messy to split them up and I think it will result in a whole lot of merge conflicts.

So I would much rather just keep to 1 version and fix all problems in here untill it can get merged.

I will stop adding new features and keep the PR static from now on, only fixing things based on feedback and small bug fixes.

starkillerOG commented:
@allenporter thanks for the review, any more suggestions?
@JPHutchins would you also be willing to take a look?

JPHutchins commented on homeassistant/components/motion_blinds/__init__.py:
```diff
@@ -31,8 +77,26 @@ async def async_setup_entry(
+    )
+
+    if len(hass.data[DOMAIN]) == 1:
+        # start listining for local pushes (only once)
```
typo: listining -> listening

starkillerOG replied:
thanks, fixed

starkillerOG committed:
fix spelling

starkillerOG commented:
@JPHutchins any more suggestions?

JPHutchins commented:
I don't know enough about the component lifecycle and event loop to make further suggestions - I'll keep watching the PR so that I can learn more!

allenporter commented:
> > One suggestion I have might be to break up the 5 different changes in the PR into separate smaller PRs that can be reviewed a bit faster, and independently.
> 
> I had thought of that, but the diffrent parts relay on diffrent versions of the upstream library and have verious other small bug fixes in them, so I think it will get a bit messy to split them up and I think it will result in a whole lot of merge conflicts.
> 
> So I would much rather just keep to 1 version and fix all problems in here untill it can get merged.
> 
> I will stop adding new features and keep the PR static from now on, only fixing things based on feedback and small bug fixes.

That sounds fine to me.  For most users multicast should do just fine, and all commands should go through the gateway, so definitely don't want to argue for very much special treatment here.

starkillerOG commented:
@bdraco would you be willing to revieuw/merge this PR?

allenporter commented:
> > One suggestion I have might be to break up the 5 different changes in the PR into separate smaller PRs that can be reviewed a bit faster, and independently.
> 
> I had thought of that, but the diffrent parts relay on diffrent versions of the upstream library and have verious other small bug fixes in them, so I think it will get a bit messy to split them up and I think it will result in a whole lot of merge conflicts.
> 
> So I would much rather just keep to 1 version and fix all problems in here untill it can get merged.
> 
> I will stop adding new features and keep the PR static from now on, only fixing things based on feedback and small bug fixes.

I'd be super happy to review / merge this in a few smaller chunks since it would be easier to fully review and feel confident i'm not missing something.  I'm having trouble setting aside a large chunk of time to make sure i'm not missing anything keeping it all straight in my head.

starkillerOG commented:
@allenporter I will see if I have some time to split up the PR soon. Will do my best

starkillerOG commented:
I broke up this big PR in a first smaller piece that only has the local push improvements in it.
The new PR is here: https://github.com/home-assistant/core/pull/44391
@allenporter could you revieuw/merge that smaller PR?

starkillerOG commented:
The first piece about local pushes has been merged.
The second piece about TDBU control is here: https://github.com/home-assistant/core/pull/44500

starkillerOG commented:
I will close this PR since the first 2 pieces are already merged in smaller PR's.
I will open up a PR for the last piece in a bit.

starkillerOG commented:
The last piece of this PR about discovery can be found here: https://github.com/home-assistant/core/pull/44615
